File: backend/run_tests.py
#!/usr/bin/env python3
"""
Simple test runner to demonstrate test execution.
Note: This requires pytest and other test dependencies to be installed.
"""

import subprocess
import sys
import os

def run_command(cmd, description):
    """Run a command and return success status."""
    print(f"\n{'='*60}")
    print(f"Running: {description}")
    print(f"Command: {cmd}")
    print('='*60)
    
    try:
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
        if result.returncode == 0:
            print(f"✅ {description} - PASSED")
            if result.stdout:
                print(result.stdout)
            return True
        else:
            print(f"❌ {description} - FAILED")
            if result.stderr:
                print("Error output:")
                print(result.stderr)
            if result.stdout:
                print("Standard output:")
                print(result.stdout)
            return False
    except Exception as e:
        print(f"❌ {description} - ERROR: {str(e)}")
        return False

def main():
    """Run all tests with proper reporting."""
    print("🧪 WhatsApp Scheduler - Backend Test Suite")
    print("==========================================")
    
    # Check if we're in the backend directory
    if not os.path.exists("tests"):
        print("❌ Error: 'tests' directory not found. Please run from the backend directory.")
        sys.exit(1)
    
    # List of test commands to run
    test_commands = [
        ("pytest tests/test_whatsapp_service.py -v", "WhatsApp Service Tests"),
        ("pytest tests/test_scheduling.py -v", "Scheduling Tests"),
        ("pytest tests/test_api_endpoints.py -v", "API Endpoint Tests"),
        ("pytest tests/ -v --cov=app --cov-report=term-missing", "All Tests with Coverage"),
    ]
    
    all_passed = True
    
    print("\n📋 Test Plan:")
    for cmd, desc in test_commands:
        print(f"  - {desc}")
    
    # Run each test suite
    for cmd, description in test_commands:
        if not run_command(cmd, description):
            all_passed = False
    
    # Summary
    print("\n" + "="*60)
    print("📊 Test Summary")
    print("="*60)
    
    if all_passed:
        print("✅ All tests passed successfully!")
        sys.exit(0)
    else:
        print("❌ Some tests failed. Please check the output above.")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: frontend/run_tests.js
#!/usr/bin/env node
/**
 * Simple test runner for frontend tests
 */

const { exec } = require('child_process');
const path = require('path');

// Colors for output
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  bold: '\x1b[1m'
};

function runCommand(command, description) {
  return new Promise((resolve) => {
    console.log(`\n${colors.blue}${'='.repeat(60)}${colors.reset}`);
    console.log(`${colors.bold}Running: ${description}${colors.reset}`);
    console.log(`Command: ${command}`);
    console.log(`${colors.blue}${'='.repeat(60)}${colors.reset}`);

    const process = exec(command, (error, stdout, stderr) => {
      if (error) {
        console.log(`${colors.red}❌ ${description} - FAILED${colors.reset}`);
        console.error(`Error: ${error.message}`);
        if (stderr) console.error(`Stderr: ${stderr}`);
        resolve(false);
      } else {
        console.log(`${colors.green}✅ ${description} - PASSED${colors.reset}`);
        if (stdout) console.log(stdout);
        resolve(true);
      }
    });

    // Stream output in real-time
    process.stdout?.on('data', (data) => {
      process.stdout.write(data);
    });

    process.stderr?.on('data', (data) => {
      process.stderr.write(data);
    });
  });
}

async function main() {
  console.log(`${colors.bold}🧪 WhatsApp Scheduler - Frontend Test Suite${colors.reset}`);
  console.log('==========================================');

  const testCommands = [
    { 
      command: 'npm test -- --watchAll=false --testMatch="**/MessageScheduler.test.tsx"', 
      description: 'MessageScheduler Component Tests' 
    },
    { 
      command: 'npm test -- --watchAll=false', 
      description: 'All Unit Tests' 
    },
    { 
      command: 'npm run test:coverage', 
      description: 'Tests with Coverage Report' 
    }
  ];

  console.log(`\n${colors.yellow}📋 Test Plan:${colors.reset}`);
  testCommands.forEach(({ description }) => {
    console.log(`  - ${description}`);
  });

  let allPassed = true;

  // Run each test suite
  for (const { command, description } of testCommands) {
    const passed = await runCommand(command, description);
    if (!passed) {
      allPassed = false;
      break; // Stop on first failure for faster feedback
    }
  }

  // Summary
  console.log(`\n${colors.blue}${'='.repeat(60)}${colors.reset}`);
  console.log(`${colors.bold}📊 Test Summary${colors.reset}`);
  console.log(`${colors.blue}${'='.repeat(60)}${colors.reset}`);

  if (allPassed) {
    console.log(`${colors.green}✅ All tests passed successfully!${colors.reset}`);
    process.exit(0);
  } else {
    console.log(`${colors.red}❌ Some tests failed. Please check the output above.${colors.reset}`);
    process.exit(1);
  }
}

// Run the test suite
main().catch((error) => {
  console.error(`${colors.red}Fatal error: ${error.message}${colors.reset}`);
  process.exit(1);
});